polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -253,13 +259,33 @@ async fn run<Context>(
 	ctx.spawn_blocking("pvf-validation-host", task.boxed())?;
 
 	let mut tasks = FuturesUnordered::new();
+	let mut session_index: Option<SessionIndex> = None;
+	let mut is_next_session_authority = false;
+	// PVF host won't prepare the same code hash twice, so here we just avoid extra communication
+	let mut already_prepared_code_hashes = HashSet::new();
 
 	loop {
 		loop {
 			futures::select! {
 				comm = ctx.recv().fuse() => {
 					match comm {
-						Ok(FromOrchestra::Signal(OverseerSignal::ActiveLeaves(_))) => {},
+						Ok(FromOrchestra::Signal(OverseerSignal::ActiveLeaves(update))) => {
+							let mut sender = ctx.sender().clone();
+
+							let Some(leaf) = update.activated else { continue };
+							let new_session_index = new_session_index(&mut sender, session_index, leaf.hash).await;
+							if new_session_index.is_some() {
+								session_index = new_session_index;
+								already_prepared_code_hashes.clear();
```

Comment:
Why clearing here? Once we prepare it should be there until restart.

Reply:
I thought the same, but chances are that by this point, the artifact might have been pruned because it had been unused for more than 24 hours. So clearing here is more foolproof than not clearing, but some better solution may also be considered.

Reply:
I supposed we should start from scratch on every new session. It’s just a mechanism to not overload the pvf host with unnecessary checking.